### PR TITLE
dbft: Shut down Geth gracefully

### LINKED ISSUE #32: Shut down Geth gracefully
https://github.com/bane-labs/go-ethereum/blob/fd059099a26002fa053bac14a6ce42c7d671a338/Makefile#L142
The database can be corrupted if we kill the geth instance.

image

We can use `SIGINT` or some other options.

### PR DESCRIPTION
**Pull request by chenquanyu**
Close #32 Shut down Geth gracefully. This issue is caused by the DBFT ```Close()``` method reads channel from not started ```eventLoop()```.

### CODE REVIEW

**Comment by nicolegys** [comment]
Works well, except that bootnode process cannot be killed by `kill -2`

**Comment by chenquanyu** [comment]
> Works well, except that bootnode process cannot be killed by `kill -2`
> 
> 

The ```bootnode``` is only a dev tool of geth, and the kill signal is not handled in this tool. We will use bootstrap node instead of ```bootnode``` in test and prod environment.

**Comment by nicolegys** [comment]
OK for me.

File: Makefile [hunk 1]
```diff
@@ -222,11 +222,11 @@ privnet_init_seven: privnet_clean
 
 privnet_nodes_stop:
 	@echo "Killing nodes processes"
-	@killall -w -v -9 geth || :
+	@killall -v -2 geth || :
```
Comment (roman-khimov) [inline Makefile]: I'd rather do `-INT` here and keep `-w` to have nodes stopped after the command. This fix is not related to `consensus/dbft` changes anyway, can you make it a separate commit?

Comment (chenquanyu) [reply Makefile]: > I'd rather do `-INT` here and keep `-w` to have nodes stopped after the command. This fix is not related to `consensus/dbft` changes anyway, can you make it a separate commit?

We can use ```killall -INT``` here and I removed the ```-w``` option because it is not compatible in Mac OS. I modify make file in this PR because ```killall -INT``` will work after this fix of issue #32 .

Comment (roman-khimov) [reply Makefile]: `-w` can be emulated with a loop (`killall` returns 1 when it doesn't actually send any signal), the problem is predictable behavior after `make privnet_nodes_stop`, without `-w` you have to check for the status somehow to ensure nodes have stopped.

PR is OK, I'm more about commit structure. I know NGD uses squashing for PRs as a policy, but I'd really prefer to see something like #79 with each commit solving exactly one problem and overall PR being merged unsquashed. This leaves much more relevant details about changes for the future. Not critical here, the PR is rather small, but notice for example our `-w` discussion, nothing in commits says that it's a problem for Mac. In ideal world I'd do `git log` and see this change as a separate commit with an explanation of _why_ it is needed. So, just a suggestion.

Comment (chenquanyu) [reply Makefile]: OK, I will fix ```-w``` in a separate commit.

**Review by AnnaShaleva — COMMENTED** [review]
Please, organize commits wrt https://github.com/bane-labs/go-ethereum/pull/85#discussion_r1445707505. So in this PR we'd expect to see two commits: the first one is Makefile change with appropriate commit message and commit body containing the reason of changes and `Close #85.` line. And the second one is `consensus/dbft/dbft.go` fix with appropriate commit message.

PR content itself LGTM.

**Comment by AnnaShaleva** [comment]
@chenquanyu, see also https://github.com/nspcc-dev/.github/blob/master/git.md#commits, please.

**Review by AnnaShaleva — COMMENTED** [review]
Sorry for nitpicking, but please, split into two commits like described in https://github.com/bane-labs/go-ethereum/pull/85#pullrequestreview-1813583676:
The first commit's header/message is expected to be like that:
```
Makefile: shut down Geth privnet gracefully

Close #32.
```
The second one is something like this:
```
dbft: properly shutdown DBFT engine

Close `quit` channel only if DBFT engine was started.
```

**Review by AnnaShaleva — APPROVED** [review]
